numba/byteflow.py

```diff
@@ -0,0 +1,1139 @@
+                      new_stack)
+        self._stack = new_stack
+
+    def get_top_block(self, kind):
+        for bs in reversed(self._blockstack):
+            if bs['kind'] == kind:
+                return bs
+
+    def get_varname(self, inst):
+        """Get referenced variable name from the oparg
+        """
+        return self._bytecode.co_varnames[inst.arg]
+
+    def terminate(self):
+        """Mark block as terminated
+        """
+        self._terminated = True
+
+    def fork(self, pc, npop=0):
+        """Fork the state
+        """
+        assert 0 <= npop <= len(self._stack)
+        nstack = len(self._stack) - npop
+        stack = self._stack[:nstack]
+        self._outedges.append(Edge(pc=pc, stack=stack,
+                                   blockstack=tuple(self._blockstack)))
+        self.terminate()
+
+    def split_new_block(self):
+        """Split the state
+        """
+        self.fork(pc=self._pc)
+
+    def get_outgoing_states(self):
+        """Get states for each outgoing edges
+        """
+        # Should only call once
+        assert not self._outgoing_phis
+        ret = []
+        for edge in self._outedges:
+            state = State(bytecode=self._bytecode, pc=edge.pc,
+                          nstack=len(edge.stack), blockstack=edge.blockstack)
+            ret.append(state)
+            # Map outgoing_phis
+            for phi, i in state._phis.items():
+                self._outgoing_phis[phi] = edge.stack[i]
+        return ret
+
+
+Edge = namedtuple("Edge", ["pc", "stack", "blockstack"])
+
+
+class AdaptDFA(object):
+    """Adapt Flow to the old DFA class expected by Interpreter
+    """
+    def __init__(self, flow):
+        self._flow = flow
+
+    @property
+    def infos(self):
+        return self._flow.block_infos
+
+
+AdaptBlockInfo = namedtuple("AdaptBlockInfo", ["insts", "outgoing_phis"])
+
+
+def adapt_state_infos(state):
+    return AdaptBlockInfo(
+        insts=tuple(state.instructions), outgoing_phis=state.outgoing_phis
+    )
+
+
+def _flatten_inst_regs(iterable):
+    """Flatten an iterable of registers used in an instruction
+    """
+    for item in iterable:
+        if isinstance(item, str):
+            yield item
+        elif isinstance(item, (tuple, list)):
+            for x in _flatten_inst_regs(item):
+                yield x
+
+
+class AdaptCFA(object):
+    """Adapt Flow to the old CFA class expected by Interpreter
+    """
+    def __init__(self, flow):
+        self._flow = flow
+        self._blocks = {}
+        for offset, blockinfo in flow.block_infos.items():
+            self._blocks[offset] = AdaptCFBlock(blockinfo, offset)
+        backbone = self._flow.cfgraph.backbone()
+
+        graph = flow.cfgraph
+        # Find backbone
+        backbone = graph.backbone()
+        # Filter out in loop blocks (Assuming no other cyclic control blocks)
+        # This is to unavoid variable defined in loops to be considered as
```

Comment:
Perhaps `This is to avoid variables defined in loops being considered as function scope`?